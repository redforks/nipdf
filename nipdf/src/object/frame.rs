use super::{Dictionary, Object, ObjectValueError, XRefSection};
use crate::file::EncryptDict;
use nipdf_macro::pdf_object;
use std::num::NonZeroU32;

/// Document id, two binary string.
pub struct DocId(pub Box<[u8]>, pub Box<[u8]>);

impl TryFrom<&Object> for DocId {
    type Error = ObjectValueError;

    fn try_from(o: &Object) -> Result<Self, Self::Error> {
        let arr = o.arr()?;
        if arr.len() != 2 {
            return Err(ObjectValueError::UnexpectedType);
        }

        Ok(Self(
            arr[0].as_byte_string()?.into(),
            arr[1].as_byte_string()?.into(),
        ))
    }
}

#[pdf_object(())]
pub trait TrailerDictTrait {
    fn size(&self) -> i32;
    fn prev(&self) -> Option<i32>;
    fn root(&self) -> Option<NonZeroU32>;
    #[nested]
    fn encrypt(&self) -> Option<EncryptDict<'a, 'b>>;
    #[key("ID")]
    #[try_from]
    fn id(&self) -> Option<DocId>;
}

#[derive(Debug, Clone)]
/// Frame contains things like xref, trailer, caused by incremental update. See [FrameSet]
pub struct Frame {
    pub xref_pos: u32,
    pub trailer: Dictionary,
    pub xref_section: XRefSection,
}

impl Frame {
    pub fn new(xref_pos: u32, trailer: Dictionary, xref_section: XRefSection) -> Self {
        Self {
            xref_pos,
            trailer,
            xref_section,
        }
    }
}

pub type FrameSet<'a> = Vec<Frame>;
