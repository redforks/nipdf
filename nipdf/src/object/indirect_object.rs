use super::{Object, ObjectId};
use std::num::NonZeroU32;

#[derive(Debug, Clone, PartialEq)]
pub struct IndirectObject(ObjectId, Object);

impl<'a> IndirectObject {
    pub fn new(id: NonZeroU32, generation: u16, object: Object) -> Self {
        Self(ObjectId::new(id, generation), object)
    }

    pub fn id(&self) -> ObjectId {
        self.0
    }

    pub fn object(&self) -> &Object {
        &self.1
    }

    pub fn take(self) -> Object {
        self.1
    }
}
