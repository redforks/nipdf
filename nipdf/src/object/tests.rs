use super::*;
use crate::file::{ObjectResolver, XRefTable};
use prescript::name;
use prescript_macro::name;
use static_assertions::assert_impl_all;
use test_case::test_case;

#[test_case("", b"()"; "empty")]
#[test_case("a", b"(a)"; "single character")]
#[test_case("a(", b"(a\\()"; "left square")]
#[test_case("a)", b"(a\\))"; "right square")]
#[test_case("ab", b"(a\\\nb)"; "escape next \\n")]
#[test_case("ab", b"(a\\\rb)"; "escape next \\r")]
#[test_case("ab", b"(a\\\r\nb)"; "escape next \\r\\n")]
#[test_case("ab", b"(a\\\n\rb)"; "escape next \\n\\r")]
#[test_case("a\nb", b"(a\\\n\nb)"; "escape one next new line")]
#[test_case("a\nb", b"(a\nb)"; "normal new line")]
#[test_case("a\nb", b"(a\rb)"; "normal \\n new line")]
#[test_case("a\nb", b"(a\r\nb)"; "normal \\r\\n new line")]
#[test_case("a\nb", b"(a\n\rb)"; "normal \\n\\r new line")]
#[test_case("\x05a", b"(\\5a)"; "oct 1")]
#[test_case("\x05a", b"(\\05a)"; "oct 2")]
#[test_case("\x05a", b"(\\005a)"; "oct 3")]
fn literal_string_decoded(exp: &str, buf: impl AsRef<[u8]>) {
    assert_eq!(LiteralString::new(buf.as_ref()).as_str(), exp);
}

#[test_case(b"", b"<>" ; "empty")]
#[test_case(b"\x90\x1f\xa3", b"<901FA3>"; "not empty")]
#[test_case(b"\x90\x1f\xa0", b"<901FA>"; "append 0 if odd")]
#[test_case(b"\x90\x1f\xa0", b"<90 1F\tA>"; "ignore whitespace")]
fn hex_string_decoded(exp: impl AsRef<[u8]>, buf: impl AsRef<[u8]>) {
    assert_eq!(HexString::new(buf.as_ref()).as_bytes(), exp.as_ref());
}

#[test_case(Ok(10), "unknown"; "not exist use default value")]
#[test_case(Ok(1), "a"; "id exist, and is int")]
#[test_case(Err(ObjectValueError::UnexpectedType), "b"; "id exist, but not int")]
fn dict_get_int(exp: Result<i32, ObjectValueError>, id: &str) {
    let mut d = Dictionary::default();
    d.set(name!("a"), 1i32);
    d.set(name!("b"), "(2)");

    assert_eq!(exp, d.get_int(name(id), 10));
}

#[test_case(Object::LiteralString(LiteralString::new(b"(foo)")), "(foo)"; "literal string")]
#[test_case(Object::HexString(HexString::new(b"<901FA3>")), "<901FA3>"; "hex string")]
#[test_case(Object::Name(name!("foo")), "/foo"; "name")]
fn buf_or_str_to_object<'a>(exp: Object, s: &'a str) {
    assert_eq!(exp, Object::from(s.as_bytes()));
    assert_eq!(exp, Object::from(s));
}

#[test]
fn dict_get_bool() {
    let mut d = Dictionary::default();
    d.set(name!("a"), true);
    d.set(name!("b"), true);
    d.set(name!("c"), 1i32);

    assert_eq!(Ok(true), d.get_bool(name!("a"), false));
    assert_eq!(Ok(true), d.get_bool(name!("b"), true));
    assert_eq!(
        Err(ObjectValueError::UnexpectedType),
        d.get_bool(name!("c"), false)
    );
    assert_eq!(Ok(false), d.get_bool(name!("d"), false));
}

#[test]
fn dict_get_name() {
    let mut d = Dictionary::default();
    d.set(name!("a"), "/foo");
    d.set(name!("b"), "/bar");
    d.set(name!("c"), 1i32);

    assert_eq!(Ok(Some(name!("foo"))), d.get_name(name!("a")));
    assert_eq!(Ok(Some(name!("bar"))), d.get_name(name!("b")));
    assert_eq!(
        Err(ObjectValueError::UnexpectedType),
        d.get_name(name!("c"))
    );
    assert_eq!(Ok(None), d.get_name(name!("d")));
}

#[test]
fn equal_schema_type_validator() {
    let checker = EqualTypeValueChecker::new(name!("Page"));
    assert!(!checker.check(None));
    assert!(!checker.check(Some(name!("blah"))));
    assert!(checker.check(Some(name!("Page"))));
}

#[test]
fn value_type_validator() {
    let validator = ValueTypeValidator::new(
        NameTypeValueGetter::new(name!("Type")),
        EqualTypeValueChecker::new(name!("Page")) as EqualTypeValueChecker<Name>,
    );
    assert_impl_all!(
        ValueTypeValidator<NameTypeValueGetter, EqualTypeValueChecker<Name>>: TypeValidator
    );

    let mut d = Dictionary::default();
    d.set(name!("a"), "/foo");

    assert_eq!(
        Err(ObjectValueError::DictSchemaUnExpectedType(
            "/Type: Page".into()
        )),
        validator.valid(&d)
    );
}

#[test]
fn option_value_type_validator() {
    let checker = EqualTypeValueChecker::new(name!("Page")).option();
    assert_impl_all!(OptionTypeValueChecker<EqualTypeValueChecker<Name>>: TypeValueCheck<Name>);

    assert!(checker.check(None));
    assert!(!checker.check(Some(name!("blah"))));
    assert!(checker.check(Some(name!("Page"))));
}

#[test]
fn one_of_type_value_checker() {
    let checker = OneOfTypeValueChecker::new(vec![name!("Page"), name!("Pages")]);
    let schema_type = <OneOfTypeValueChecker<Name> as TypeValueCheck<Name>>::schema_type(&checker);
    assert_eq!("/Page|/Pages", &schema_type);

    assert!(!checker.check(None::<Name>));
    assert!(!checker.check(Some(name!("blah"))));
    assert!(checker.check(Some(name!("Page"))));
    assert!(checker.check(Some(name!("Pages"))));
}

#[test_case(None => Vec::<u32>::new())]
#[test_case(Some(&[]) => Vec::<u32>::new())]
#[test_case(Some(&[1, 2]) => vec![1, 2])]
fn schema_ref_id_arr(ids: Option<&[u32]>) -> Vec<u32> {
    let mut d = Dictionary::new();
    if let Some(ids) = ids {
        let ids: Array = ids.iter().map(|id| Object::new_ref(*id)).collect();
        d.insert(name!("ids"), ids.into());
    }
    let xref = XRefTable::empty();
    let resolver = ObjectResolver::empty(&xref);
    let d = SchemaDict::new(&d, &resolver, ()).unwrap();
    d.ref_id_arr(name!("ids"))
        .unwrap()
        .into_iter()
        .map(|id| id.get())
        .collect()
}

#[cfg(feature = "pretty")]
#[test_case(Object::Null => "null")]
#[test_case(true => "true")]
#[test_case(1i32 => "1")]
#[test_case(1.0f32 => "1.0")]
#[test_case(LiteralString::new(b"(foo)") => "(foo)")]
#[test_case(HexString::new(b"<901FA3>") => "<901fa3>")]
#[test_case("/foo" => "/foo"; "Name")]
#[test_case(vec![] => "[]"; "empty array")]
#[test_case(vec![Object::Null] => "[null]"; "array with null")]
#[test_case(vec![1i32.into(), 2i32.into()] => "[1 2]"; "array with two int")]
#[test_case(15u32 => "15 0 R"; "reference")]
#[test_case(Dictionary::new() => "<<>>"; "empty dict")]
#[test_case([(name!("a"), true.into())].into_iter().collect::<Dictionary>() => "<</a true>>"; "dict with one entry")]
#[test_case([(name!("a"), true.into()), (name!("b"), false.into())].into_iter().collect::<Dictionary>() => "<</a true /b false>>"; "dict with two entries")]
fn pretty_print(o: impl Into<Object>) -> String {
    let o = o.into();
    let s = o.to_doc().pretty(20).to_string();
    s
}

#[test]
fn f32_arr_try_from_object() {
    let arr = vec![1.0f32.into(), 2.0f32.into()];
    let o = Object::Array(arr);
    let arr2: [f32; 2] = (&o).try_into().unwrap();
    assert_eq!([1.0f32, 2.0f32], arr2);
}
