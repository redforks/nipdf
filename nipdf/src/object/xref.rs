use super::RuntimeObjectId;

#[derive(Debug, Clone, PartialEq, Copy)]
pub struct FilePos(u32, u16, bool); // offset, generation, is_used

impl FilePos {
    pub fn new(offset: u32, generation: u16, is_used: bool) -> Self {
        Self(offset, generation, is_used)
    }

    pub fn offset(&self) -> u32 {
        self.0
    }

    pub fn generation(&self) -> u16 {
        self.1
    }

    pub fn is_used(&self) -> bool {
        self.2
    }
}

#[derive(Debug, Clone, PartialEq, Copy)]
pub enum Entry {
    /// The object stored directly in file as object
    InFile(FilePos),
    /// The object stored in a stream, (stream_object_id, idx_of_object_in_stream)
    InStream(RuntimeObjectId, u16),
}

impl Entry {
    pub fn in_file(offset: u32, generation: u16, is_used: bool) -> Self {
        Self::InFile(FilePos::new(offset, generation, is_used))
    }

    pub fn in_stream(stream_object_id: RuntimeObjectId, idx_of_object_in_stream: u16) -> Self {
        Self::InStream(stream_object_id, idx_of_object_in_stream)
    }

    pub fn is_used(self) -> bool {
        match self {
            Entry::InFile(pos) => pos.is_used(),
            Entry::InStream(_, _) => true,
        }
    }
}

pub type Section = Vec<(u32, Entry)>;
