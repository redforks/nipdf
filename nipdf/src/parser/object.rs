use std::{borrow::Cow, num::NonZeroU32, str::from_utf8, str::FromStr};

use nom::{
    branch::alt,
    bytes::complete::{escaped, is_not, tag, take, take_till, take_while},
    character::{
        complete::{anychar, line_ending, multispace1, u16, u32},
        is_hex_digit,
    },
    combinator::{map, not, opt, peek, recognize, value},
    error::{ErrorKind, FromExternalError},
    multi::{many0, many0_count},
    number::complete::float,
    sequence::{delimited, preceded, separated_pair, terminated, tuple},
};

use crate::object::{
    Array, Dictionary, HexString, IndirectObject, LiteralString, Name, Object, ObjectValueError,
    Reference, Stream,
};

use super::{whitespace_or_comment, ws, ws_prefixed, ws_terminated, ParseError, ParseResult};

/// Unwrap the result of nom parser to a *normal* result.
pub fn unwrap_parse_result<'a, T: 'a>(obj: ParseResult<'a, T>) -> Result<T, ParseError<'a>> {
    match obj {
        Ok((_, obj)) => Ok(obj),
        Err(nom::Err::Incomplete(_)) => unreachable!(),
        Err(nom::Err::Error(e)) | Err(nom::Err::Failure(e)) => Err(e),
    }
}

pub fn parse_object(buf: &[u8]) -> ParseResult<Object> {
    let null = value(Object::Null, tag(b"null"));
    let true_parser = value(Object::Bool(true), tag(b"true"));
    let false_parser = value(Object::Bool(false), tag(b"false"));
    let number_parser = map(recognize(float), |s| {
        if memchr::memchr(b'.', s).is_some() {
            let s = from_utf8(s).unwrap();
            f32::from_str(s).map(Object::Number).unwrap()
        } else {
            let s = from_utf8(s).unwrap();
            i32::from_str(s)
                .map(Object::Integer)
                .unwrap_or_else(|_| Object::Number(f32::from_str(s).unwrap()))
        }
    });

    fn parse_quoted_string(input: &[u8]) -> ParseResult<&[u8]> {
        let esc = escaped(is_not("\\()"), '\\', anychar);
        let inner_parser = alt((esc, parse_quoted_string));
        let mut parser = recognize(delimited(tag(b"("), many0_count(inner_parser), tag(b")")));
        parser(input)
    }
    let parse_quoted_string = map(parse_quoted_string, |s| {
        Object::LiteralString(LiteralString::new(s))
    });
    let parse_hex_string = map(
        recognize(delimited(
            tag(b"<"),
            take_while(|c| is_hex_digit(c) || c.is_ascii_whitespace()),
            tag(b">"),
        )),
        |buf| Object::HexString(HexString::new(buf)),
    );

    alt((
        map(parse_name, Object::Name),
        parse_quoted_string,
        map(parse_dict, Object::Dictionary),
        map(parse_array, Object::Array),
        parse_hex_string,
        null,
        true_parser,
        false_parser,
        map(parse_reference, Object::Reference),
        number_parser,
    ))(buf)
}

/// Return `Err(ObjectValueError::InvalidNameForma)` if the name is not a valid PDF name encoding,
/// not two hex char after `#`.
fn normalize_name(buf: &[u8]) -> Result<Cow<[u8]>, ObjectValueError> {
    fn next_hex_char(iter: &mut impl Iterator<Item = u8>) -> Option<u8> {
        let mut result = 0;
        for _ in 0..2 {
            if let Some(c) = iter.next() {
                result <<= 4;
                result |= match c {
                    b'0'..=b'9' => c - b'0',
                    b'a'..=b'f' => c - b'a' + 10,
                    b'A'..=b'F' => c - b'A' + 10,
                    _ => return None,
                };
            } else {
                return None;
            }
        }
        Some(result)
    }

    let s = &buf[1..];
    if s.iter().copied().any(|b| b == b'#') {
        let mut result = Vec::with_capacity(s.len());
        let mut iter = s.iter().copied();
        while let Some(next) = iter.next() {
            if next == b'#' {
                if let Some(c) = next_hex_char(&mut iter) {
                    result.push(c);
                } else {
                    return Err(ObjectValueError::InvalidNameFormat);
                }
            } else {
                result.push(next);
            }
        }
        Ok(Cow::Owned(result))
    } else {
        Ok(Cow::Borrowed(s))
    }
}

fn parse_name(input: &[u8]) -> ParseResult<Name> {
    let (input, buf) = recognize(preceded(
        tag(b"/"),
        take_till(|c: u8| {
            c.is_ascii_whitespace()
                || c == b'['
                || c == b'<'
                || c == b'('
                || c == b'/'
                || c == b'>'
                || c == b']'
        }),
    ))(input)?;
    let name = normalize_name(buf)
        .map_err(|e| nom::Err::Error(ParseError::from_external_error(input, ErrorKind::Fail, e)))
        .map(Name)?;
    Ok((input, name))
}

pub fn parse_array(input: &[u8]) -> ParseResult<Array> {
    delimited(
        ws(tag(b"[")),
        many0(ws(parse_object)),
        ws_terminated(tag(b"]")),
    )(input)
}

pub fn parse_dict(input: &[u8]) -> ParseResult<Dictionary> {
    map(
        delimited(
            ws(tag(b"<<".as_slice())),
            many0(tuple((parse_name, ws(parse_object)))),
            ws_terminated(tag(b">>")),
        ),
        |v| v.into_iter().collect(),
    )(input)
}

fn parse_object_and_stream(input: &[u8]) -> ParseResult<Object> {
    let (input, o) = parse_object(input)?;
    match o {
        Object::Dictionary(d) => {
            let (input, begin_stream) = opt(delimited(
                whitespace_or_comment,
                tag(b"stream"),
                line_ending,
            ))(input)?;
            if begin_stream.is_some() {
                if let Some(Object::Integer(l)) = d.get(&b"Length"[..]) {
                    let (input, data) = take(*l as usize)(input)?;
                    let (input, _) = preceded(opt(line_ending), tag(b"endstream"))(input)?;
                    Ok((input, Object::Stream(Stream::new(d, data))))
                } else {
                    Ok((input, Object::Stream(Stream::new(d, input))))
                }
            } else {
                Ok((input, Object::Dictionary(d)))
            }
        }
        _ => Ok((input, o)),
    }
}

/// Parse Stream that its length not ref_id(no need to resolve), return
/// &[u8] that just after endstream
fn parse_stream_that_length_not_ref_id(input: &[u8]) -> ParseResult<Stream> {
    let (input, (dict, _)) = tuple((
        parse_dict,
        delimited(whitespace_or_comment, tag(b"stream"), line_ending),
    ))(input)?;
    let (input, data) = take(dict.get(&b"Length"[..]).unwrap().as_int().unwrap() as usize)(input)?;
    let (input, _) = preceded(line_ending, tag(b"endstream"))(input)?;
    Ok((input, Stream::new(dict, data)))
}

pub fn parse_indirected_object(input: &[u8]) -> ParseResult<'_, IndirectObject<'_>> {
    let (input, (id, gen)) = separated_pair(u32, multispace1, u16)(input)?;
    let (input, obj) = preceded(ws(tag(b"obj")), parse_object_and_stream)(input)?;
    let (input, _) = opt(ws_prefixed(tag("endobj")))(input)?;
    Ok((
        input,
        IndirectObject::new(NonZeroU32::new(id).unwrap(), gen, obj),
    ))
}

/// Parse stream wrapped in indirect object tag,
/// different from `parse_indirected_object()`, buf will after the end of `endobj`
pub fn parse_indirected_stream(input: &[u8]) -> ParseResult<(NonZeroU32, Stream)> {
    let (input, (id, _)) = separated_pair(u32, multispace1, u16)(input)?;
    let (input, stream) = preceded(ws(tag(b"obj")), parse_stream_that_length_not_ref_id)(input)?;
    let (input, _) = ws(tag(b"endobj"))(input)?;
    Ok((input, (NonZeroU32::new(id).unwrap(), stream)))
}

fn parse_reference(input: &[u8]) -> ParseResult<Reference> {
    let (input, (id, gen)) = terminated(
        separated_pair(u32, multispace1, u16),
        // `not(peek(tag("G")))` to detect `RG` graphics operation,
        // such as `0 1 0 RG` is a valid graphics operation,
        // if not check `not(peek(tag("G")))`, the sequence will parsed to
        // Integer(0), Reference(1 0)
        delimited(whitespace_or_comment, tag(b"R"), not(peek(tag("G")))),
    )(input)?;
    Ok((input, Reference::new_u32(id, gen)))
}

#[cfg(test)]
mod tests;
