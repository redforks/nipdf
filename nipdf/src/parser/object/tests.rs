use super::*;
use crate::object::HexString;
use test_case::test_case;

#[test_case(Object::Null, "null"; "null")]
#[test_case(true, "true")]
#[test_case(false, "false")]
#[test_case(123, "123"; "integer")]
#[test_case(-123, "-123"; "negative integer")]
#[test_case(123, "+123"; "integer prefixed with +")]
#[test_case(32488685, "32488685"; "integer can not cast from float")]
#[test_case(4294967296f32, "4294967296"; "integer out of range")]
#[test_case(123.12, "123.12"; "number")]
#[test_case(-123.12, "-123.12"; "negative number")]
#[test_case(123.12, "+123.12"; "number prefixed with +")]
#[test_case(4.0, "4.0"; "number end with dot")]
#[test_case(-0.002, "-.002"; "number start with dot")]
#[test_case(LiteralString::new(b"()"), "()"; "empty literal string")]
#[test_case(LiteralString::new(b"(5\\()"), "(5\\()"; "escaped )")]
#[test_case(LiteralString::new(b"(5\\\\)"), "(5\\\\)"; "escaped back slash")]
#[test_case(LiteralString::new(b"(a(foo))"), "(a(foo))"; "nested quoted string")]
#[test_case(LiteralString::new(b"(a
b)"), "(a
b)"; "literal string contains new line")]
#[test_case(LiteralString::new(b"(*!&}^%)"), "(*!&}^%)"; "literal string contains special characters")]
#[test_case(LiteralString::new(b"(\\)\\()"), "(\\)\\()"; "literal string contains escape")]
#[test_case(LiteralString::new(b"(\\333\\n)"), "(\\333\\n)")]
#[test_case(HexString::new(b"<>"), "<>")]
#[test_case(HexString::new(b"<12A>"), "<12A>")]
#[test_case(HexString::new(b"<12 A\t3>"), "<12 A\t3>"; "contains whitespace")]
#[test_case(name!(""), "/"; "empty name")]
#[test_case(name!("foo"), "/foo"; "name")]
fn test_parse_simple_objects(exp: impl Into<Object>, buf: &'static str) {
    let o = parse_object(buf.as_bytes()).unwrap();
    assert_eq!((b"".as_slice(), exp.into()), o);
}

#[test_case(vec![], "[]"; "empty array")]
#[test_case(vec![], "[ \t]"; "empty array 2")]
#[test_case(vec![Object::Null], "[null]"; "array with null")]
#[test_case(vec![Object::Array(vec![Object::Null].into())], "[[null]]"; "nested array with null")]
#[test_case(vec![name!("foo").into()], "[/foo]"; "name value")]
fn test_parse_array(exp: Vec<Object>, buf: &'static str) {
    assert_eq!(
        (b"".as_slice(), exp.into()),
        parse_array(buf.as_bytes()).unwrap()
    );
}

#[test_case(b"<< >>", "empty dict")]
#[test_case(b"<<>>", "empty dict 2")]
#[test_case(b"<< /Type /Catalog >>", "dict with one entry")]
#[test_case(b"<</Inner<<>>>>", "nested")]
#[test_case(b"<</id[]>>", "empty array")]
#[test_case(b"<</id()>>", "string value")]
#[test_case(b"<</id/Value>>", "name value")]
#[test_case(b"<</id/>>", "empty name value")]
#[test_case(b"<<//id>>", "empty name key")]
#[test_case(b"<</id<<>>>>", "nested empty dict")]
fn test_parse_dict(buf: impl AsRef<[u8]>, name: &str) {
    insta::assert_debug_snapshot!(name, parse_dict(buf.as_ref()).unwrap());
}

#[test_case(
    b"1 0 obj
null
endobj",
    "null"
)]
fn test_parse_indirected_object(buf: impl AsRef<[u8]>, name: &str) {
    insta::assert_debug_snapshot!(name, parse_indirect_object(buf.as_ref()).unwrap());
}

#[test_case(b"1 0 R", "simple")]
fn test_parse_reference(buf: impl AsRef<[u8]>, name: &str) {
    insta::assert_debug_snapshot!(name, parse_reference(buf.as_ref()).unwrap());
}

#[test_case("foo", b"/foo")]
#[test_case("a#b", b"/a#23b")]
#[test_case("Ab", b"/#41#62")]
fn name_normalize(exp: impl AsRef<str>, name: impl AsRef<[u8]>) {
    assert_eq!(normalize_name(name.as_ref()).unwrap(), exp.as_ref());
}

#[test]
fn test_parse_object_and_stream() {
    // length is int
    let buf = br#"<</Length 4>>
stream
1234
endstream
"#;
    let (input, o) = parse_object_and_stream(buf).unwrap();
    assert_eq!(input, b"\n");
    let (_, start, length) = o.right().unwrap();
    assert_eq!(21, start);
    assert_eq!(Some(NonZeroU32::new(4).unwrap()), length);

    // length is ref
    let buf = br#"<</Length 1 0 R>>
stream
blah
endstream
"#;
    let (input, o) = parse_object_and_stream(buf).unwrap();
    assert_eq!(input[0], b'b');
    assert!(input.len() > 4);
    let (_, start, length) = o.right().unwrap();
    assert_eq!(25, start);
    assert_eq!(None, length);

    // endstream precede with cr
    let buf = b"<</Length 4>>
stream
1234\rendstream
";
    let (input, o) = parse_object_and_stream(buf).unwrap();
    assert_eq!(input, b"\n");
    let (_, start, length) = o.right().unwrap();
    assert_eq!(21, start);
    assert_eq!(Some(NonZeroU32::new(4).unwrap()), length);

    // length is 0
    let buf = b"<</Length 0>>
stream
endstream
";
    let (input, o) = parse_object_and_stream(buf).unwrap();
    assert_eq!(input, b"\n");
    let (_, start, length) = o.right().unwrap();
    assert_eq!(21, start);
    assert_eq!(None, length);
}
