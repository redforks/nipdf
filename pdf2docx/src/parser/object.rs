use std::borrow::Cow;

use nom::{
    branch::alt,
    bytes::{
        complete::{escaped, is_not, tag, take_till, take_until, take_while},
        streaming::take,
    },
    character::{
        complete::{anychar, crlf, multispace1, u16, u32},
        is_hex_digit,
    },
    combinator::{map, not, opt, peek, recognize, value},
    multi::{many0, many0_count},
    number::complete::float,
    sequence::{delimited, preceded, separated_pair, terminated, tuple},
};
use num::cast;

use crate::object::{
    Array, Dictionary, HexString, IndirectObject, LiteralString, Name, Object, ObjectValueError,
    Reference, Stream,
};

use super::{whitespace_or_comment, ws, ws_prefixed, ws_terminated, ParseError, ParseResult};

/// Unwrap the result of nom parser to a *normal* result.
pub fn unwrap_parse_result<'a, T: 'a>(obj: ParseResult<'a, T>) -> Result<T, ParseError<'a>> {
    match obj {
        Ok((_, obj)) => Ok(obj),
        Err(nom::Err::Incomplete(_)) => unreachable!(),
        Err(nom::Err::Error(e)) | Err(nom::Err::Failure(e)) => Err(e),
    }
}

pub fn parse_object(buf: &[u8]) -> ParseResult<Object> {
    let null = value(Object::Null, tag(b"null"));
    let true_parser = value(Object::Bool(true), tag(b"true"));
    let false_parser = value(Object::Bool(false), tag(b"false"));
    let number_parser = map(float, |v| {
        if let Some(i) = cast(v) {
            if v.fract() == 0.0 {
                Object::Integer(i)
            } else {
                Object::Number(v)
            }
        } else {
            Object::Number(v)
        }
    });

    fn parse_quoted_string(input: &[u8]) -> ParseResult<&[u8]> {
        let esc = escaped(is_not("\\()"), '\\', anychar);
        let inner_parser = alt((esc, parse_quoted_string));
        let mut parser = recognize(delimited(tag(b"("), many0_count(inner_parser), tag(b")")));
        parser(input)
    }
    let parse_quoted_string = map(parse_quoted_string, |s| {
        Object::LiteralString(LiteralString::new(s))
    });
    let parse_hex_string = map(
        recognize(delimited(
            tag(b"<"),
            take_while(|c| is_hex_digit(c) || c.is_ascii_whitespace()),
            tag(b">"),
        )),
        |buf| Object::HexString(HexString::new(buf)),
    );

    alt((
        map(parse_name, Object::Name),
        parse_quoted_string,
        map(parse_dict, Object::Dictionary),
        map(parse_array, Object::Array),
        parse_hex_string,
        null,
        true_parser,
        false_parser,
        map(parse_reference, Object::Reference),
        number_parser,
    ))(buf)
}

/// Return `Err(ObjectValueError::InvalidNameForma)` if the name is not a valid PDF name encoding,
/// not two hex char after `#`.
fn normalize_name(buf: &[u8]) -> Result<Cow<[u8]>, ObjectValueError> {
    fn next_hex_char(iter: &mut impl Iterator<Item = u8>) -> Option<u8> {
        let mut result = 0;
        for _ in 0..2 {
            if let Some(c) = iter.next() {
                result <<= 4;
                result |= match c {
                    b'0'..=b'9' => c - b'0',
                    b'a'..=b'f' => c - b'a' + 10,
                    b'A'..=b'F' => c - b'A' + 10,
                    _ => return None,
                };
            } else {
                return None;
            }
        }
        Some(result)
    }

    let s = &buf[1..];
    if s.iter().copied().any(|b| b == b'#') {
        let mut result = Vec::with_capacity(s.len());
        let mut iter = s.iter().copied();
        while let Some(next) = iter.next() {
            if next == b'#' {
                if let Some(c) = next_hex_char(&mut iter) {
                    result.push(c);
                } else {
                    return Err(ObjectValueError::InvalidNameFormat);
                }
            } else {
                result.push(next);
            }
        }
        Ok(Cow::Owned(result))
    } else {
        Ok(Cow::Borrowed(s))
    }
}

fn parse_name(input: &[u8]) -> ParseResult<Name> {
    let (input, buf) = recognize(preceded(
        tag(b"/"),
        take_till(|c: u8| {
            c.is_ascii_whitespace()
                || c == b'['
                || c == b'<'
                || c == b'('
                || c == b'/'
                || c == b'>'
                || c == b']'
        }),
    ))(input)?;
    let name = normalize_name(buf)
        .map_err(|_| nom::Err::Error(ParseError::InvalidNameFormat))
        .map(Name)?;
    Ok((input, name))
}

pub fn parse_array(input: &[u8]) -> ParseResult<Array> {
    delimited(
        ws(tag(b"[")),
        many0(ws(parse_object)),
        ws_terminated(tag(b"]")),
    )(input)
}

pub fn parse_dict(input: &[u8]) -> ParseResult<Dictionary> {
    map(
        delimited(
            ws(tag(b"<<".as_slice())),
            many0(tuple((parse_name, ws(parse_object)))),
            ws_terminated(tag(b">>")),
        ),
        |v| v.into_iter().collect(),
    )(input)
}

pub fn parse_stream_content(input: &[u8], stream_len: u32) -> ParseResult<&[u8]> {
    delimited(
        tuple((ws_prefixed(tag(b"stream")), alt((crlf, tag(b"\n"))))),
        take(stream_len),
        ws(tag(b"endstream")),
    )(input)
}

fn parse_object_and_stream(input: &[u8]) -> ParseResult<Object> {
    let (input, o) = parse_object(input)?;
    let (input, buf) = match o {
        Object::Dictionary(ref d) => {
            let Some(len) = d.get(&Name::borrowed(b"Length")) else {
                return Ok((input, o));
            };
            match len {
                Object::Integer(len) => opt(delimited(
                    tuple((ws_prefixed(tag(b"stream")), alt((crlf, tag(b"\n"))))),
                    take(*len as u32),
                    ws(tag(b"endstream")),
                ))(input)?,
                Object::Reference(_) => {
                    let (input, inner) = peek(opt(ws_prefixed(tag(b"stream"))))(input)?;
                    return match inner {
                        None => Ok((input, o)),
                        Some(_) => {
                            let (input, _) =
                                take_until::<&[u8], &[u8], ParseError>(b"endobj")(input)?;
                            Ok((input, Object::LaterResolveStream(d.clone())))
                        }
                    };
                }
                _ => return Ok((input, o)),
            }
        }
        _ => return Ok((input, o)),
    };
    match buf {
        Some(buf) => match o {
            Object::Dictionary(d) => Ok((input, Object::Stream(Stream(d, buf)))),
            _ => unreachable!(),
        },
        None => Ok((input, o)),
    }
}

pub fn parse_indirected_object(input: &[u8]) -> ParseResult<IndirectObject> {
    let (input, (id, gen)) = separated_pair(u32, multispace1, u16)(input)?;
    let (input, obj) =
        delimited(ws(tag(b"obj")), parse_object_and_stream, ws(tag(b"endobj")))(input)?;
    Ok((input, IndirectObject::new(id, gen, obj)))
}

fn parse_reference(input: &[u8]) -> ParseResult<Reference> {
    let (input, (id, gen)) = terminated(
        separated_pair(u32, multispace1, u16),
        // `not(peek(tag("G")))` to detect `RG` graphics operation,
        // such as `0 1 0 RG` is a valid graphics operation,
        // if not check `not(peek(tag("G")))`, the sequence will parsed to
        // Integer(0), Reference(1 0)
        delimited(whitespace_or_comment, tag(b"R"), not(peek(tag("G")))),
    )(input)?;
    Ok((input, Reference::new(id, gen)))
}

#[cfg(test)]
mod tests;
