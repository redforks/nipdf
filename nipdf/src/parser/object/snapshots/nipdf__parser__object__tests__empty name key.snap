---
source: nipdf/src/parser/object/tests.rs
expression: parse_dict(buf.as_ref()).unwrap()
---
(
    [],
    Dictionary(
        {
            name!(): Name(
                name!(id),
            ),
        },
    ),
)
